pigpen-cascading/src/main/clojure/pigpen/cascading.clj

```diff
@@ -41,335 +68,228 @@
+                (zip [p (-> fold :projections next)
+                      {:keys [pipe]} ancestors
+                      key keys]
+                  (if (some-> p :expr :udf #{:fold})
+                    (let [context (pr-str `'{:projections ~(prepare-projections [p])})
+                          group-fields (cfields [key])
+                          arg-fields (cfields (get-in p [:expr :args]))]
+                      (PigPenAggregateBy. context pipe group-fields arg-fields))
+                    pipe)))]
+    (CoGroup. (str id) pipes join-keys Fields/NONE (BufferJoin.))))
+
+(s/defmethod command->flowdef :join
+  [{:keys [id keys fields join-types opts]} :- m/Join
+   ancestors
+   _]
+  (let [pipes (->> ancestors
+                (map :pipe)
+                (pipe-array))
+        joiner (->> join-types
+                 (map (comp boolean #{:required}))
+                 (boolean-array)
+                 (MixedJoin.))
+        join-keys (group-key-cfields keys (:join-nils opts))]
+    (CoGroup. (str id) pipes join-keys (cfields fields) joiner)))
+
+(s/defmethod command->flowdef :generate
+  [{:keys [id projections fields]} :- m/Mapcat
+   [{:keys [^Pipe pipe ancestor]}]
+   _]
+  (let [context (pr-str `'{:ancestor ~(update-in ancestor [:opts] dissoc :tap)
+                           :projections ~(prepare-projections projections)
+                           :fields ~fields})]
+    (case (:type ancestor)
+      :reduce
+      (Every. pipe (ReduceBuffer. context (cfields fields)) Fields/RESULTS)
+
+      (:group :group-fold)
+      (Every. pipe (GroupBuffer. context (cfields fields)) Fields/RESULTS)
+
+      ;else
+      (-> (Pipe. (str id) pipe)
+        (Each. (PigPenFunction. context (cfields fields)))))))
+
+(s/defmethod command->flowdef :distinct
+  [{:keys [fields]} :- m/Distinct
+   [{:keys [^Pipe pipe ancestor]}]
+   _]
+  (-> pipe
+    (Unique. Fields/ALL)
+    (Rename. (cfields (:fields ancestor)) (cfields fields))))
+
+(s/defmethod command->flowdef :limit
+  [{:keys [n fields]} :- m/Take
+   [{:keys [^Pipe pipe ancestor]}]
+   _]
+  (-> pipe
+    (Each. (Limit. n))
+    (Rename. (cfields (:fields ancestor)) (cfields fields))))
+
+(s/defmethod command->flowdef :sample
+  [{:keys [p fields]} :- m/Sample
+   [{:keys [^Pipe pipe ancestor]}]
+   _]
+  (-> pipe
+    (Each. (Sample. p))
+    (Rename. (cfields (:fields ancestor)) (cfields fields))))
+
+(s/defmethod command->flowdef :union
+  [{:keys [fields]} :- m/Concat
+   ancestors
+   _]
+  (->> ancestors
+    (map (fn [{:keys [^Pipe pipe ancestor]}]
+           (Rename. pipe (cfields (:fields ancestor)) (cfields fields))))
+    (into-array Pipe)
+    (Merge.)))
+
+(s/defmethod command->flowdef :order
+  [{:keys [key comp fields]} :- m/Sort
+   [{:keys [^Pipe pipe ancestor]}]
+   _]
+  (let [reverse-order? (= :desc comp)]
+    (-> pipe
+      (GroupBy. Fields/NONE (cfields [key]) reverse-order?)
+      ;; TODO is there a way to rename and select a single field at the same time?
```

Comment:
I think this can be achieved by passing the new name as an argument to Identity, so that the explicit rename is not needed.